sky/serve/autoscalers.py

```diff
@@ -179,44 +193,61 @@ def collect_request_information(
 
         # Convert to requests per second.
-        num_requests_per_second = len(
-            self.request_timestamps) / self.qps_window_size
-        target_num_replicas = math.ceil(num_requests_per_second /
-                                        self.target_qps_per_replica)
-        target_num_replicas = max(self.min_replicas,
-                                  min(self.max_replicas, target_num_replicas))
-        logger.info(f'Requests per second: {num_requests_per_second}, '
-                    'Current/proposed target number of replicas: '
-                    f'{self.target_num_replicas}/{target_num_replicas}')
-
-        if not self.bootstrap_done or self.target_num_replicas == 0:
-            self.bootstrap_done = True
-            return target_num_replicas
+        target_num_replicas = self._cal_target_num_replicas_based_on_qps()
+        old_target_num_replicas = self.target_num_replicas
+
+        # Faster scale up when there is no replica.
+        if self.target_num_replicas == 0:
+            self.target_num_replicas = target_num_replicas
         elif target_num_replicas > self.target_num_replicas:
             self.upscale_counter += 1
             self.downscale_counter = 0
             if self.upscale_counter >= self.scale_up_consecutive_periods:
                 self.upscale_counter = 0
-                return target_num_replicas
+                self.target_num_replicas = target_num_replicas
         elif target_num_replicas < self.target_num_replicas:
             self.downscale_counter += 1
             self.upscale_counter = 0
             if self.downscale_counter >= self.scale_down_consecutive_periods:
                 self.downscale_counter = 0
-                return target_num_replicas
+                self.target_num_replicas = target_num_replicas
         else:
             self.upscale_counter = self.downscale_counter = 0
-        return self.target_num_replicas
+
+        num_requests_per_second = len(
+            self.request_timestamps) / self.qps_window_size
+        logger.info(
+            f'Requests per second: {num_requests_per_second}. '
+            f'Current target number of replicas: {old_target_num_replicas}. '
+            f'Final target number of replicas: {self.target_num_replicas}. '
+            f'Upscale counter: {self.upscale_counter}/'
+            f'{self.scale_up_consecutive_periods}. '
+            f'Downscale counter: {self.downscale_counter}/'
+            f'{self.scale_down_consecutive_periods}')
+
+    @classmethod
+    def select_replicas_to_scale_down(
+            cls, num_limit: int,
+            replica_infos: Iterable['replica_managers.ReplicaInfo']
+    ) -> List[int]:
+
+        status_order = serve_state.ReplicaStatus.scale_down_decision_order()
+        replica_infos_sorted = sorted(
+            replica_infos,
+            key=lambda info: status_order.index(info.status)
+            # Terminate undefined status.
+            if info.status in status_order else -1)
```

Comment:
Added a todo. I think this seems to be a minor addition that could be added in the future

Reply:
`status_order` only includes `[cls.PENDING, cls.PROVISIONING, cls.STARTING, cls.READY]`
This is to consider other states.

Reply:
Ahh, I see. I guess we can either have all the status in the status_order or list those missing status in the comment instead of calling them undefined : )

Reply:
Added!